pkg/controller/direct/apigee/envgroup_controller.go

```diff
@@ -0,0 +1,274 @@
+	desired              *api.GoogleCloudApigeeV1EnvironmentGroup
+	actual               *api.GoogleCloudApigeeV1EnvironmentGroup
+	apigeeEnvgroupClient *apigeeEnvgroupClient
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return true means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ApigeeEnvgroup", "name", a.id)
+
+	envgroup, err := a.apigeeEnvgroupClient.featureClient.Get(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ApigeeEnvgroup %q: %w", a.id, err)
+	}
+
+	a.actual = envgroup
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ApigeeEnvgroup", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	op, err := a.apigeeEnvgroupClient.featureClient.Create(a.id.Parent().String(), a.desired).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("creating ApigeeEnvgroup %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	if err := a.waitForOp(ctx, op); err != nil {
+		return fmt.Errorf("waiting for ApigeeEnvgroup %s creation: %w", a.id, err)
+	}
+
+	created, err := a.apigeeEnvgroupClient.featureClient.Get(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting created ApigeeEnvgroup: %w", err)
+	}
+
+	log.V(2).Info("successfully created ApigeeEnvgroup", "ApigeeEnvgroup", created)
+
+	status := &krm.ApigeeEnvgroupStatus{}
+	status.ObservedState = ApigeeEnvgroupObservedState_FromApi(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("patching ApigeeEnvgroup", a.fullyQualifiedName())
+
+	updateMask := fieldmaskpb.FieldMask{}
+
+	// Sorts the Hostname lists so that the comparison is deterministic
+	if !reflect.DeepEqual(asSortedCopy(a.desired.Hostnames), asSortedCopy(a.actual.Hostnames)) {
+		log.V(2).Info("change detected: hostnames")
+		updateMask.Paths = append(updateMask.Paths, "hostnames")
+	}
+
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
```

Comment:
Here, we need to update status also. This is a recent change to the controller template. When adopting existing resources, it might be the case that no fields actually change, but we still need to update the status to print output only fields. Example: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/680a1271c0a6ca2b16ceb972820de8d923a0cd20/pkg/controller/direct/workstations/workstation_controller.go#L211-L217

Reply:
Done